src/java.net.http/share/classes/java/net/http/HttpClient.java

```diff
@@ -730,4 +745,126 @@ public enum Redirect {
     public WebSocket.Builder newWebSocketBuilder() {
         throw new UnsupportedOperationException();
     }
+
+    /**
+     * Initiates an orderly shutdown in which previously submitted
+     * operations are completed, but no new request will be accepted.
+     * Invocation has no additional effect if already shut down.
+     *
+     * <p>This method does not wait for previously submitted request
+     * to complete execution.  Use {@link #awaitTermination awaitTermination}
+     * to do that.
+     *
+     * @implSpec
+     * The default implementation of this method does nothing. Subclasses should
+     * override this method to implement the appropriate behavior.
+     * See the API Note on {@linkplain ##closing closing}.
+     *
+     * @since 21
+     */
+    public void shutdown() { }
+
+    /**
+     * Blocks until all operations have completed execution after a shutdown
+     * request, or the timeout occurs, or the current thread is
```

Comment:
If I understand this correctly, this method is expected to be called after `shutdown()` or `shutdownNow()` has been called. If so, then should this method throw an `IllegalStateException` if it gets called before any of those shutdown methods are called?

Reply:
Would reword this sentence to:

> Blocks until all operations have completed execution after a shutdown request, or the {@code duration} passes, or the current thread is {@linkplain Thread#interrupt() interrupted}, whichever happens first.

Reply:
> should this method throw an `IllegalStateException` if it gets called before any of those shutdown methods are called?

ExecutorService doesn't do that, and I don't think we should here either. The actual implementation of the method just joins the selector manager thread. There is no requirement that shutdown() and awaitTermination() are performed in the same thread, and no technical reason to prevent waiting before shutdown (especially if shutdown is triggered by some other thread).

> Do you think we could reword this sentence to, to have necessary pointers to duration and thread interruption:

Good idea.